// @material-ui/styles/styled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _withStyles = _interopRequireDefault(require("./withStyles"));

// styled-components's API removes the mapping between components and styles.
// Using components as a low-level styling construct can be simpler.
function styled(Component) {
  var componentCreator = function componentCreator(style, options) {
    function StyledComponent(props) {
      var classes = props.classes,
          className = props.className,
          other = (0, _objectWithoutProperties2.default)(props, ["classes", "className"]);
      return _react.default.createElement(Component, (0, _extends2.default)({
        className: (0, _classnames.default)(classes.root, className)
      }, other));
    }

    StyledComponent.propTypes = process.env.NODE_ENV !== "production" ? {
      classes: _propTypes.default.object.isRequired,
      className: _propTypes.default.string,
      theme: _propTypes.default.object
    } : {};
    var styles = typeof style === 'function' ? function (theme) {
      return {
        root: style(theme)
      };
    } : {
      root: style
    };
    return (0, _withStyles.default)(styles, options)(StyledComponent);
  };

  return componentCreator;
}

var _default = styled;
exports.default = _default;
