// @material-ui/styles/ThemeProvider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ThemeContext = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _warning = _interopRequireDefault(require("warning"));

var _utils = require("@material-ui/utils");

/* eslint-disable no-underscore-dangle */
// In order to have self-supporting components, we rely on default theme when not provided.
_utils.ponyfillGlobal.__MUI_DEFAULT_THEME__ = _utils.ponyfillGlobal.__MUI_DEFAULT_THEME__ || {};
var defaultTheme = _utils.ponyfillGlobal.__MUI_DEFAULT_THEME__;

var ThemeContext = _react.default.createContext(defaultTheme); // To support composition of theme.


exports.ThemeContext = ThemeContext;

function mergeOuterLocalTheme(outerTheme, localTheme) {
  if (typeof localTheme === 'function') {
    process.env.NODE_ENV !== "production" ? (0, _warning.default)(outerTheme, ['Material-UI: you are providing a theme function property ' + 'to the ThemeProvider component:', '<ThemeProvider theme={outerTheme => outerTheme} />', 'However, no outer theme is present.', 'Make sure a theme is already injected higher in the React tree ' + 'or provide a theme object.'].join('\n')) : void 0;
    var mergedTheme = localTheme(outerTheme);
    process.env.NODE_ENV !== "production" ? (0, _warning.default)(mergedTheme, 'Material-UI: return an object from your theme function, i.e. theme={() => ({})}!') : void 0;
    return mergedTheme;
  }

  return (0, _extends2.default)({}, outerTheme, localTheme);
}
/**
 * This component takes a `theme` property.
 * It makes the `theme` available down the React tree thanks to React context.
 * This component should preferably be used at **the root of your component tree**.
 */


function ThemeProvider(props) {
  var children = props.children,
      localTheme = props.theme;
  return _react.default.createElement(ThemeContext.Consumer, null, function (outerTheme) {
    var theme = outerTheme === defaultTheme ? localTheme : mergeOuterLocalTheme(outerTheme, localTheme);
    return _react.default.createElement(ThemeContext.Provider, {
      value: theme
    }, children);
  });
}

ThemeProvider.propTypes = process.env.NODE_ENV !== "production" ? {
  /**
   * You can wrap a node.
   */
  children: _propTypes.default.node.isRequired,

  /**
   * A theme object.
   */
  theme: _propTypes.default.oneOfType([_propTypes.default.object, _propTypes.default.func]).isRequired
} : {};
ThemeProvider.propTypes = process.env.NODE_ENV !== "production" ? (0, _utils.exactProp)(ThemeProvider.propTypes) : {};
var _default = ThemeProvider;
exports.default = _default;
